src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs

```diff
@@ -102,16 +102,42 @@ public async Task SpecializeAsync()
         {
             _logger.LogInformation("Starting language worker channel specialization");
             _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);
+
             ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
             if (_workerRuntime != null && languageWorkerChannel != null)
             {
-                _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
-                await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                if (UsePlaceholderChannel(_workerRuntime))
+                {
+                    _logger.LogDebug("Loading environment variables for runtime: {runtime}", _workerRuntime);
+                    await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                }
+                else
+                {
+                    _logger.LogDebug("Shutting down placeholder worker. Worker is not compatible for runtime: {runtime}", _workerRuntime);
+                    // If we need to allow file edits, we should shutdown the webhost channel on specialization.
+                    await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
+                }
             }
+
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderChannel(string workerRuntime)
+        {
```

Comment:
simplify - this method needs to return true if it is ever called with runtime not set. FunctionDispatcher will  shutdown placeholder channels after jobhost initializes

```
internal bool UsePlaceholderChannel(string workerRuntime)
        {
            if (!string.IsNullOrEmpty(workerRuntime))
            {
                // Special case: node apps must be read-only to use the placeholder mode channel
                // TODO: Remove special casing when resolving https://github.com/Azure/azure-functions-host/issues/4534
                if (string.Equals(workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
                {
                    return _environment.FileSystemIsReadOnly();
                }
            }
            return true;
        }
```

Reply:
If workerRuntime isn't set, then we won't be able to use the channel that was initialized in placeholder mode because we won't know which channel we should be using (ex: to specialize/sending environment variables to). So I think that if workerRuntime is not set, then we should not use any channel that was started in placeholder mode (and if we were to support workerRuntime that is not set, there are more changes needed to support that, like sending environment variables to all worker channels started in placeholder mode). 

The check for null or empty workerRuntime isn't necessary either because before it is called, we make sure that workerRuntime is not null or empty. I want to keep that logic there though in case some other method chooses to call `UsePlaceholderChannel` without the same check, and I think in this case the result should be "false"

Reply:
unfortunately, if workerruntime is not set, there is no action at webhost level. Placeholder channel needs to be disposed at jobhost level. To avoid others calling into this method, make this method private to webhostchannelmanager?

Reply:
Sounds good! I think I made it internal for unit tests, but we have tests on SpecializeAsync too so I can make this private or just put it into the logic for SpecializeAsync.

Reply:
yes either one works! thanks!